### ISSUE TITLE
how to compute the (bures) wasserstein distance between gaussians of different dimensions

### ISSUE THREAD

**Issue by ttsesm** [ott-jax/ott#490]
Hi,

I am trying to compute the (bures) wasserstein distance between gaussians of different dimensions. Mainly I was checking on the tutorial shown [here](https://ott-jax.readthedocs.io/en/latest/tutorials/gromov_wasserstein_multiomics.html) where they try to find couplings of datasets with different sizes but of the same dimentions. However, my dataset differs a bit from the one used in the example thus any help would be appreciated.

For example I have ten 2D gaussians and ten 3D gaussians on which I would like to find their coupling (in practice the coupling is 1-to-1, meaning the first 2D gaussian corresponds to the first 3D gaussian and so on since it is a test case). For each gaussian set I have the corresponding means and covariances as follows:

**2D gaussians:**
```
Covariances
ct = np.array([[[0.00049392, 0.00018833],
        [0.00018833, 0.00007765]],
       [[0.00037171, 0.00021303],
        [0.00021303, 0.00012497]],
       [[0.00004997, 0.00002954],
        [0.00002954, 0.00012977]],
       [[0.00005111, 0.00000863],
        [0.00000863, 0.00002055]],
       [[0.00022878, 0.00014253],
        [0.00014253, 0.00023158]],
       [[0.00010538, 0.00007804],
        [0.00007804, 0.00030141]],
       [[0.00001159, 0.00000829],
        [0.00000829, 0.00000615]],
       [[0.00007754, 0.00002211],
        [0.00002211, 0.00011025]],
       [[0.00070302, 0.00060664],
        [0.00060664, 0.00052477]],
       [[0.00109267, 0.00122162],
        [0.00122162, 0.00137384]]], dtype=np.float32)

Means
mt = np.array([[-0.34575066, -0.05927338],
       [-0.16067392, -0.15656751],
       [ 0.0884882 ,  0.7348885 ],
       [-0.09218572,  0.05595953],
       [ 0.21521865,  0.38351792],
       [-0.22240582,  0.30064186],
       [-0.0520223 , -0.18196055],
       [-0.4030493 ,  0.22060394],
       [-0.17406262, -0.09412329],
       [ 0.3939966 ,  0.21849754]], dtype=np.float32)
```
**3D gaussians:**
```
Covariances
Cs = np.array([[[ 0.00055704,  0.0002181 ,  0.00020697],
        [ 0.0002181 ,  0.00026268,  0.00017932],
        [ 0.00020697,  0.00017932,  0.00013152]],

       [[ 0.00029736,  0.00036497, -0.00025849],
        [ 0.00036497,  0.0005222 , -0.00034746],
        [-0.00025849, -0.00034746,  0.00028704]],

       [[ 0.00013889, -0.00002427,  0.00009434],
        [-0.00002427,  0.000106  , -0.00010998],
        [ 0.00009434, -0.00010998,  0.00020017]],

       [[ 0.00013997,  0.0000322 , -0.00003177],
        [ 0.0000322 ,  0.00013866,  0.0000159 ],
        [-0.00003177,  0.0000159 ,  0.00008226]],

       [[ 0.00015876,  0.00015987,  0.00013226],
        [ 0.00015987,  0.00046588, -0.0000928 ],
        [ 0.00013226, -0.0000928 ,  0.00045325]],

       [[ 0.00014214, -0.00009106,  0.00025811],
        [-0.00009106,  0.00042213,  0.00002785],
        [ 0.00025811,  0.00002785,  0.00058036]],

       [[ 0.00003409, -0.00000352,  0.00000672],
        [-0.00000352,  0.00000058, -0.00000025],
        [ 0.00000672, -0.00000025,  0.00000226]],

       [[ 0.00025886, -0.00007174, -0.00000447],
        [-0.00007174,  0.00020796,  0.00000071],
        [-0.00000447,  0.00000071,  0.00017473]],

       [[ 0.00000316, -0.00008337, -0.00002749],
        [-0.00008337,  0.00292826,  0.00008949],
        [-0.00002749,  0.00008949,  0.00106939]],

       [[ 0.00080745, -0.00143929, -0.00031194],
        [-0.00143929,  0.00750104,  0.00179461],
        [-0.00031194,  0.00179461,  0.0016364 ]]], dtype=np.float32)

Means
Ms = np.array([[ 0.17411666,  0.5035763 ,  0.42941484],
       [ 0.5251834 , -0.27657598, -0.18892948],
       [-0.5073861 ,  0.46957424, -0.90752286],
       [ 0.56422734, -0.4835634 , -0.91173404],
       [-0.5370846 ,  0.22988465, -0.2466668 ],
       [ 0.04097709,  0.48167574, -0.23595962],
       [ 0.49995327, -0.5364788 , -0.3129182 ],
       [ 0.4078352 ,  0.46962914, -0.3338474 ],
       [ 0.5331596 , -0.24272342, -0.30946448],
       [-0.57427835, -0.20426048, -0.28931147]], dtype=np.float32)
```
In the example they have two 2D datasets of different dimensions but it is not straight forward how to adapt it for my case scenario. It should be possible to do that, right?

The complete problem, I am describing it in the discussion [here](https://github.com/ott-jax/ott/discussions/489) just in case.

**Comment by michalk8** [comment]
Hi @ttsesm , thanks for your interest. Below is a way of computing the GW using the [Bures cost](https://ott-jax.readthedocs.io/en/latest/_autosummary/ott.geometry.costs.Bures.html#ott.geometry.costs.Bures):
```python
import jax
import jax.numpy as jnp

from ott.geometry import pointcloud, geometry, costs
from ott.solvers.quadratic import gromov_wasserstein
from ott.problems.quadratic import quadratic_problem


# load the data
mt = ...
ct = ...
Ms = ...
Cs = ...


cost_fn_2d = costs.Bures(2)  # 2 dim
cost_fn_3d = costs.Bures(3)  # 3 dims

xx = jax.vmap(costs.mean_and_cov_to_x, in_axes=[0, 0, None])(mt, ct, 2)
yy = jax.vmap(costs.mean_and_cov_to_x, in_axes=[0, 0, None])(Ms, Cs, 3)
print(xx.shape, yy.shape)  # ((10, 6), (10, 12))

materialize = False  # might be faster when True
geom_xx = pointcloud.PointCloud(xx, cost_fn=cost_fn_2d)
geom_yy = pointcloud.PointCloud(yy, cost_fn=cost_fn_3d)
if materialize:
    c_xx = geom_xx.cost_matrix
    c_yy = geom_yy.cost_matrix
    
    geom_xx = geometry.Geometry(c_xx)
    geom_yy = geometry.Geometry(c_yy)

epsilon = 1e-2  # select some meaningful value
solver = jax.jit(gromov_wasserstein.GromovWasserstein(epsilon=epsilon))
prob = quadratic_problem.QuadraticProblem(geom_xx, geom_yy)

out = solver(prob)

transport_matrix = out.matrix
reg_gw_cost = out.reg_gw_cost
```

If for your use-case inner dot-product cost function makes sense, in [this paper](https://arxiv.org/abs/2108.10961) (theorem 2) they define a closed-form solution for the entropic GW (balanced Gaussians only).
And in [here](https://arxiv.org/abs/2104.07970), there's a lower bound for the non-entropic case (for the L2 cost; theorem 3.1) and even closed-form solution when further restricted to Gaussian plans.

However, the two above-mentioned approaches are not implemented in OTT (might be in the future).
Hope any of this helps!

**Comment by ttsesm** [comment]
@michalk8 thanks a lot for the feedback, as well as for the papers. 
Definitely, I will have a look on them. They look quite interesting and they might come on hand for my task.

I will test your code snippet and give an update.

**Comment by ttsesm** [comment]
Hi @michalk8,

I've run the provided code snippet and the transport matrix that I am getting is the following:

![image](https://github.com/ott-jax/ott/assets/10189018/b1b1dd7b-dede-4f6f-934a-0aaf68b8e1ae)

In practice I should be getting the matches (higher values) on the diagonal since as I mentioned earlier there is a one to one correspondence between my toy example of the 2D and 3D gaussians. As you can see though only for half of them I am getting a good match.

Is there any tweaking that I could possibly apply in any of the functions that could possibly affect the ouput? For example changing the epsilon to `1e-3` seems to have an affect:

![image](https://github.com/ott-jax/ott/assets/10189018/110a5d7c-c05b-4739-9977-044ac2cf9275)

making the matching value more significant but still having some wrong matches. Maybe some parameter on the `costs.bures()`, `quadratic_problem.QuadraticProblem()` functions? In theory the two sets are mapped linearly through a projection matrix.

**Comment by michalk8** [comment]
> Is there any tweaking that I could possibly apply in any of the functions that could possibly affect the ouput? For example changing the epsilon to 1e-3 seems to have an affect:

Yes, the lower the epsilon, the crisper the coupling (also higher computational cost + risk of hitting numerical errors if too low); choosing this really depends on application.

Other important aspect is to ensure that the source/target cost matrices are of similar scale; you can do this, e.g. by changing in the above-mentioned snipper:
```python
geom_xx = pointcloud.PointCloud(xx, cost_fn=cost_fn_2d, scale_cost='mean')
geom_yy = pointcloud.PointCloud(yy, cost_fn=cost_fn_3d, scale_cost='mean')
```
This will divide the cost matrices by their respective means, below is the difference for the example data you shared:

![image](https://github.com/ott-jax/ott/assets/46717574/1dbbb5d2-be80-4c7d-916b-ee1666b52bd4)

**Comment by michalk8** [comment]
Also in https://github.com/ott-jax/ott/discussions/489, you mention:
>  Moreover, it is not necessary that all distributions in one set match with all distributions on the other set. It can be that only few ellipsoids match with some of the ellipses in the other set. I am trying to figure out whether geometrically or statistically is possible to get something based on OT.

This can be tackled by regularizing that the marginals match (using KL divergence), instead of having a constraint that needs to be satisfied.
You can play with 2 paramters $\tau_a$ and $\tau_b$ (both in $(0, 1]$; the. higher the value, the stronger the regularization, see Definition 5 [here](https://arxiv.org/abs/2211.08775).

Below is an example what you need to change in OTT to use the unbalanced formulation:
```python
tau_a = ...  # 1.0 is the default for both
tau_b = ...
prob = quadratic_problem.QuadraticProblem(geom_xx, geom_yy, tau_a=tau_a, tau_b=tau_b)
```

**Comment by ttsesm** [comment]
Thanks @michalk8. Yes, in the discussion I mention that in the final task this will be the case where not all the ellipses (2D gaussians) will correspond "perfectly" to ellipsoids (3D gaussians). Actually, only a few good matches will be enough. 

However, this is my final case scenario. In the above example here, it is an oracle test where I know exactly that each one of the ten ellipses matches to one exact ellipsoid and to which one (1 to 1, 2 to 2, 3 to 3, etc... so the output of my transport matrix should be a diagonal). Thus, I am doing that in order to experiment and basically understand how to use OT for my task, since my experience with OT is still basic.

**Comment by ttsesm** [comment]
> > Is there any tweaking that I could possibly apply in any of the functions that could possibly affect the ouput? For example changing the epsilon to 1e-3 seems to have an affect:
> 
> Yes, the lower the epsilon, the crisper the coupling (also higher computational cost + risk of hitting numerical errors if too low); choosing this really depends on application.
> 
> Other important aspect is to ensure that the source/target cost matrices are of similar scale; you can do this, e.g. by changing in the above-mentioned snipper:
> 
> ```python
> geom_xx = pointcloud.PointCloud(xx, cost_fn=cost_fn_2d, scale_cost='mean')
> geom_yy = pointcloud.PointCloud(yy, cost_fn=cost_fn_3d, scale_cost='mean')
> ```
> 
> This will divide the cost matrices by their respective means, below is the difference for the example data you shared:
> 
> ![image](https://private-user-images.githubusercontent.com/46717574/304879978-1dbbb5d2-be80-4c7d-916b-ee1666b52bd4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwOTk1MjksIm5iZiI6MTcwODA5OTIyOSwicGF0aCI6Ii80NjcxNzU3NC8zMDQ4Nzk5NzgtMWRiYmI1ZDItYmU4MC00YzdkLTkxNmItZWUxNjY2YjUyYmQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE2VDE2MDAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1N2Q2MWI2NGIwNGI4N2MxZWQyODBiMmNmZTljZTBhMzk0ZjMwNDZjMDU5NWRhYzhmNzk3ZThhNzcxZTMxNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GVN8b1LiJjdmeAZpAiZ8FopBoQ5Stvkp1kYUTBnu5OU)

@michalk8 how did you manage to get this cost matrix?

In my case, I am getting quite a different output:

```
cost_fn_2d = costs.Bures(2)  # 2 dim
cost_fn_3d = costs.Bures(3)  # 3 dims

xx = jax.vmap(costs.mean_and_cov_to_x, in_axes=[0, 0, None])(mt, ct, 2)
yy = jax.vmap(costs.mean_and_cov_to_x, in_axes=[0, 0, None])(Ms, Cs, 3)
print(xx.shape, yy.shape)  # ((10, 6), (10, 12))

geom_xx = pointcloud.PointCloud(xx, cost_fn=cost_fn_2d, scale_cost='mean')
geom_yy = pointcloud.PointCloud(yy, cost_fn=cost_fn_3d, scale_cost='mean')

c_xx = geom_xx.cost_matrix
c_yy = geom_yy.cost_matrix

plt.matshow(np.array(c_xx))
plt.matshow(np.array(c_yy))
```

`c_xx`
![image](https://github.com/ott-jax/ott/assets/10189018/e5462661-01e2-44ae-a436-0d984a1586a2)

`c_yy`
![image](https://github.com/ott-jax/ott/assets/10189018/b80305fa-0d73-4e58-9e3b-52bd666e8cb2)

Quite more noisy as you can see :thinking: 

my values seem to be within the same range though:
![image](https://github.com/ott-jax/ott/assets/10189018/a38ce361-9e70-492a-90a2-9a0ba7e10158)

Playing with the $\tau$ values didn't seem to help me much. But I guess the problem derives initially from the cost matrix.

**Comment by michalk8** [comment]
> Actually, this raises another question whether I can filter out the "good" matches by any confidence metric based on the extracted transport matrix because if I can do that then I do not really care whether all the matches are perfect or not. 

I am not aware of any such metric, but maybe @marcocuturi knows.

> @michalk8 how did you manage to get this cost matrix?

What I plotted before was the coupling, not the cost matrix, showing the effect of the scaling.

> n the above example that I've posted here with the toy sets, it is an oracle test where I know exactly that each one of the ten ellipses matches to one exact ellipsoid and also to which one (1 to 1, 2 to 2, 3 to 3, etc... so the output of my transport matrix should be a diagonal).

Ok, I see, then

> Playing with the values didn't seem to help me much. But I guess the problem derives initially from the cost matrix.

will not help at all.

**Comment by ttsesm** [comment]
> > Actually, this raises another question whether I can filter out the "good" matches by any confidence metric based on the extracted transport matrix because if I can do that then I do not really care whether all the matches are perfect or not.
> 
> I am not aware of any such metric, but maybe @marcocuturi knows.
> 

I see, thanks.
> > @michalk8 how did you manage to get this cost matrix?
> 
> What I plotted before was the coupling, not the cost matrix, showing the effect of the scaling.
> 

Interesting, thanks for the clarification. The question though, is then why I am still getting a wrong coupling output in comparison to yours:

![image](https://github.com/ott-jax/ott/assets/10189018/1033d49b-6bd5-4a62-85c1-71ae8d0ed304)
:thinking: you are setting epsilon to `1e-3` right?
> > n the above example that I've posted here with the toy sets, it is an oracle test where I know exactly that each one of the ten ellipses matches to one exact ellipsoid and also to which one (1 to 1, 2 to 2, 3 to 3, etc... so the output of my transport matrix should be a diagonal).
> 
> Ok, I see, then
> 
> > Playing with the values didn't seem to help me much. But I guess the problem derives initially from the cost matrix.
> 
> will not help at all in this toy example.
Yup, I understood that.
